[WWST-2487] Adding new ZigBee DTH for Somfy blind controller

lukhong opened the pull request.

lukhong committed:
[WWST-2487] Adding new ZigBee DTH for Somfy blind controller

lukhong commented:
Raw FP : 01 0104 0202 00 04 0000 0003 0004 0102 01 0019

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+        
+        command "pause"
+
+        fingerprint profileId: "0104", inClusters: "0000, 0003, 0004, 0102", outClusters: "0019", deviceJoinName: "SOMFY Blind Controller/eZEX", model: "E2B0-KR000Z0-HA"// SY-IoT201-BD
```
By convention the `deviceJoinName` should be the last thing in the fingerprint. Is `SOMFY Blind Controller/eZEX` what you want to show up when the device joins?

lukhong replied:
It was from the partner I didn't check it. Is the deviceJoinName to override to name in the definition for each FP?
Then in this case, can I just change the sequence like manufacturer:  , model: and delete the deviceJoinName?

tpmanley replied:
If you want to do join name translations you'll need to keep the join name on the fingerprint but otherwise it's not technically needed and the DTH name will be used as the join name. You do need to include the `model` (and in general `manufacturer` if the device supports it) since that's what is really used for identifying the device.

lukhong replied:
I will apply like that.

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+        }
+        
+        preferences {
+          input "preset", "number", title: "Default half-open position (1-100). Springs Window Fashions users should consult their manuals.", defaultValue: 50, required: false, displayDuringSetup: false
```
inconsistent whitespace

lukhong replied:
Will be fixed in the next commit

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+              attributeState "open", label:'Open', action:"close", icon:"http://www.ezex.co.kr/img/st/window_open.png", backgroundColor:"#00A0DC", nextState:"turningOff"
+              attributeState "close", label:'Closed', action:"open", icon:"http://www.ezex.co.kr/img/st/window_close.png", backgroundColor:"#ffffff", nextState:"turningOn"
+              attributeState "turningOn", label:'Opening',  action:"close", icon:"http://www.ezex.co.kr/img/st/window_open.png", backgroundColor:"#00A0DC", nextState:"turningOff"
+              attributeState "turningOff", label:'Closing', icon:"http://www.ezex.co.kr/img/st/window_close.png", backgroundColor:"#ffffff", nextState:"turningOn"  
```
these values should be:

"open", "closed"*, "partially open", "opening", and "closing"

lukhong replied:
Fixed and will be contained in the next commit.

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+            }         
+        }
+        standardTile("home", "device.level", width: 2, height: 2, decoration: "flat") {
+          state "default", label: "home", action:"presetPosition", icon:"st.Home.home2"
```
the `presetPosition` action has been moved to the "Window Shade Preset" capability. Please include it.

greens replied:
Actually you need to define that method anyway. Perhaps the device does not support this capability and this was added by mistake?

lukhong replied:
I want to know about the presetPosition clearly. Is that meaning the specific position user want to set at any time with this command? And user should be able to adjust the value of this presetPosition, right?
Then I can say this device support presetPosition because setLevel(value) is working well. But the problem is, as we know, new SmartThings app doesn't support preference feature so user cannot modify the preset value. So I just wanted to remove all presetPosition from this DTH. But if we should support classic app yet, I will add Window Shade Preset too.

tpmanley replied:
The "preset position" has to be set out of band of the DTH. I think it probably doesn't apply to this device so I'd recommend not including the `Window Shade Preset` capability.

lukhong replied:
Is that meaning the presetPosition is only for the device that can set the preset position via H/W button or its own remote controller anyway out of the DTH?

tpmanley replied:
That's correct

lukhong replied:
Thanks for the confirmation. I will remove it.

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+      zigbee.command(0x0102, 0x00)
+}
+
+def setLevel(data) {
```
add `rate=null` to the signature here

lukhong replied:
Could you explain in detail why it should have `rate=null` even though this setLevel() is a local method in this DTH. If it also needs to treat rate in the body I could guess something but just adding to the signature didn't make sense to me.

tpmanley replied:
`setLevel` is always called with two parameters when executed via the public API and other users of the command may choose to pass `rate` or not. Basically it's optional for the caller but required to be implemented. We recently updated all of our existing DTHs to add the `rate` parameter where it was missing for this reason.

lukhong replied:
When I searched many DTHS in the public, I just could find one.
Anyway should I implement the body treating `rate` , I could copy it from appengine zigbee lib. Will it be OK?

tpmanley replied:
This is where we changed `setLevel` in  existing DTHs: https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/3999. 

For the implementation just define `setLevel` as `def setLevel(value, rate = null)` and don't do anything with the `rate` parameter since the  `Go To Lift Percentage` command doesn't take a rate parameter.

lukhong replied:
Thanks for the reference, I could understand it clearly. I didn't realize there has been the `rate `in the Switch Level.  I will apply it.

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+
+def close() {
+	  log.info "LOG-MH-Shade: off()"
+      zigbee.command(0x0102, 0x01)
```
instead of using `0x0102` several times, please create:

`def getCLUSTER_WINDOW_COVERING() {return 0x0102}`

lukhong replied:
Yes, I will apply it.

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+    zigbee.onOffRefresh() + zigbee.onOffConfig()   
+}
+
+def configure() {
```
`onOffConfig` should be sent here, not in `refresh()`

lukhong replied:
it was in the configure() before but I just removed it because this device doesn't have on-off cluster.
So, can I just remove them from refresh?

tpmanley replied:
Yes, if it doesn't have the On/Off cluster then remove both `zigbee.onOffRefresh()` and `zigbee.onOffConfig()`

lukhong replied:
I will apply it.

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+
+    
+    def parseMap = zigbee.parseDescriptionAsMap(description)
+    def event = zigbee.getEvent(description)
```
`event` is never used

lukhong replied:
I will remove it.

greens commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+        }
+        
+        preferences {
+          input "preset", "number", title: "Default half-open position (1-100). Springs Window Fashions users should consult their manuals.", defaultValue: 50, required: false, displayDuringSetup: false
```
This preference is poorly copied from the Springs DTH. Additionally, there is no update method to handle this value changing. Nor is there a method that uses this value. If the device actually supports a preset position, it should be determined whether it is set via hardware manipulation of the device or if there is some zigbee command that needs to be sent.

lukhong replied:
As I commented in about the preset above, I will add after it is clear. This device cannot be set to go to the presetPosition via H/W button but available via zigbee command like below.
 zigbee.command(0x0102, 0x05, zigbee.convertToHexString(data, 2))

lukhong replied:
@greens , as @tpmanley answered above, this device doesn't seem to support presetPosition and I will remove this preference() too.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+    def event = zigbee.getEvent(description)
+    
+    def attrs = parseMap.additionalAttrs
+    if( attrs != null ) { 
```
Please check the formatting of this code block. There should be spaces after the `if` statements and consistent indentation.

lukhong replied:
I will apply it.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+    def attrs = parseMap.additionalAttrs
+    if( attrs != null ) { 
+				attrs.each { attr ->
+            if(attr.value == "64") { //open
```
Need to check the cluster and attribute before acting on the value.

lukhong replied:
applied to the patch

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+        details(["windowShade", "contPause", "ShadeLevel", "levelSliderControl","refresh"])
+    }
+}
+
```
Define the Window Covering cluster here like this:

`private getWINDOW_COVERING_CLUSTER() { 0x0102 }`

Then elsewhere in the code you can reference it like this:

```
def close() {
      zigbee.command(WINDOW_COVERING_CLUSTER, 0x01)
}
```
Just makes it more readable

lukhong replied:
I think this comment may be for other PR....

tpmanley replied:
Nope, I just messed up :) Fixed now

lukhong replied:
Yes, as @greens suggested before, I applied it in the patch. Thanks.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,120 @@
+
+// Parse incoming device messages to generate events
+def parse(String description) {
+
```
nit: remove extra newlines

lukhong replied:
OK.

lukhong committed:
Apply patch after the first review

lukhong commented:
@tpmanley , @greens 
I submitted the patch.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,118 @@
+def refresh() {
+    def cmds =  zigbee.readAttribute(CLUSTER_WINDOW_COVERING, 0x00)
+    log.info "LOG-INFO-Shade: refresh() read covering information : ${cmds}"
+}
```
`cmds` needs to be on the last line or else nothing will be sent to the device

lukhong replied:
I will add `return cmds`

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,118 @@
+    log.info "attrs - ${attrs}"
+    if ( attrs != null ) {
+        attrs.each { attr ->
+            if(parseMap.clusterInt == CLUSTER_WINDOW_COVERING && attr.attrInt == ATTRIBUTE_POSITION_LIFT) {
```
You will need to setup attribute reporting for this attribute by calling `zigbee.configureReporting` in `configure`

lukhong replied:
I am still confused about `configure()`. Would you explain more?
I thought that `sendEvent(name: "checkInterval"` is needed for the health check.
Then, what is `zigbee.configureReporting` about the attribute for?

Additional information of this device is here if it is related to this conversation. I could find the event in the `parse` at every 5 min(exactly 295 secs.) while there is no action and change.

```
5:56:58 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:52:03 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:52:03 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:47:09 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:47:09 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:42:16 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:37:21 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:32:26 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:32:26 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:27:32 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:22:38 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
5:22:38 PM: info attrs - [[attrId:0008, attrInt:8, encoding:20, value:00, isValidForDataType:true, consumedBytes:8]]
```

tpmanley replied:
It's best not to assume the device will automatically report attributes so we always explicitly configure any reporting we need using `zigbee.configureReporting`. So in this case if you want attribute 0x0008 to be reported when it changes or every five minutes you should return the following from `configure`:

`zigbee.configureReporting(CLUSTER_WINDOW_COVERING, ATTRIBUTE_POSITION_LIFT,  DataType.UINT8, 0, 600, null)`

lukhong replied:
The following question, please. This device is actually reporting this attribute automatically. So in this case, adding this command,` zigbee.configureReporting` doesn't affect anything, right? I just applied configreReporting with `min. time` as 0 and `max. time` as 60 to see if it reports more frequently. 
`zigbee.configureReporting(CLUSTER_WINDOW_COVERING, ATTRIBUTE_POSITION_LIFT, DataType.UINT8, 0, 60, null)`

But it worked as before, reporting every 5 min. So I guess it doesn't override the device's original configure report. So you just recommended to write down the `zigbee.configureReporting` as a kind of rule, not assuming automatic report. Am I correct?

greens replied:
correct

lukhong committed:
Patch after 2nd review and small changes

1. Applying for all reviewed result
2. Changing some log message
3. Remove `manufacturer` from the definition because the firmware doesn't contain this info.

lukhong commented:
@tpmanley @greens please review this patch.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,123 @@
+/**
+ *
+ *	Copyright 2019 eZEX
```
Correct me if I'm wrong but I believe this DTH is developed by SmartThings and not eZEX so I think it should say SmartThings here and also in the `definition`

lukhong replied:
It was a little unclear to me on how to determine it. At the first time, eZEX gave me the DTH that was partially working but contained some wrong implementation. I think it could changed to SmartThings. But do we need to get an agreement on it from the manufacturer? What is the standard normally?

dkirker replied:
@lukhong It would depend on where they got the DTH from. I suspect, given the coding style, from the SmartThingsPublic project.

What might be appropriate (and actually technically required, I believe, but I am not a lawyer), if eZEX started with a ST DTH and added changes, would be:

```
Copyright 2019 SmartThings
Copyright 2019 eZEX
```

Regardless (and, disclaimer again, I am not a lawyer, so take the following as my understanding of how this stuff works), if they started with a ST DTH the `Copyright 2019 SmartThings` part (or, whatever the copyright date is in the DTH they used) is mandatory.

lukhong replied:
Then I think it could be changed to SmartThings but I am not clear yet.
How was the [SDS doorlock](https://github.com/SmartThingsCommunity/SmartThingsPublic/tree/master/devicetypes/samsungsds/samsung-smart-doorlock.src) case before? It might be started from SmartThings DTH and same style.

lukhong replied:
I just changed related things to SmartThings.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,123 @@
+        standardTile("refresh", "device.refresh", inactiveLabel: false, decoration: "flat", width: 2, height: 1) {
+            state "default", label:"", action:"refresh.refresh", icon:"st.secondary.refresh"
+        }
+        valueTile("ShadeLevel", "device.level", width: 4, height: 1) {
```
nit: usually the names start with lowercase, so `shadeLevel`

dkirker replied:
I noticed this had been marked "resolved" but does not appear to be resolved?

lukhong replied:
fixed

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,123 @@
+        }
+
+        main "windowShade"
+        details(["windowShade", "contPause", "ShadeLevel", "levelSliderControl","refresh"])
```
nit: space before `"refresh"`

dkirker replied:
I noticed this had been marked "resolved" but does not appear to be resolved?

lukhong replied:
fixed

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,123 @@
+}
+
+def refresh() {
+    def cmds =  zigbee.readAttribute(CLUSTER_WINDOW_COVERING, ATTRIBUTE_POSITION_LIFT)
```
nit: extra space after `=`

dkirker replied:
I noticed this had been marked "resolved" but does not appear to be resolved? (There is an extra space after the '='.)

lukhong replied:
fixed.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,123 @@
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+    definition(name: "SOMFY Blind Controller/eZEX", namespace: "eZEX", author: "eZEX Corp", vid:"SmartThings-smartthings-Somfy_WindowShade", ocfDeviceType: "oic.d.blind") {
```
@lukhong @greens @dkirker I think this DTH could be potentially reused for other Zigbee shades. Thoughts on renaming/moving it to `Zigbee Window Shade`? We already have a `Z-Wave Window Shade`.

lukhong replied:
I think so if possible.

dkirker replied:
I'm reading over it, but I imagine that we could very easily.

tpmanley replied:
I think we should rename it then. There is nothing device specific in here so it might be useable for other similar devices.

lukhong replied:
Yeah, then I can push a commit that changes the name and the file path.
Then the vid should be changed to `generic-shade`, right? But it needs to be updated because it shows as 'dimming level' not the 'shade level'. Can @dkirker change the `generic-shade` in order to be fit to shade device?

lukhong replied:
Oh, `vid:` is working in the fingerprint! I will make the commit for this change.

dkirker replied:
`generic-shade` can be updated (and should be).

lukhong replied:
@dkirker , I want to know if you are able to update `generic-shade` and merge this PR before next Tuesday so that the certification step start during next week?

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,123 @@
+    log.info "LOG-INFO-Shade: In parse(), additionalAttrs: - ${attrs}"
+    if ( attrs != null ) {
+        attrs.each { attr ->
+            if(parseMap.clusterInt == CLUSTER_WINDOW_COVERING && attr.attrInt == ATTRIBUTE_POSITION_LIFT) {
```
nit: space after `if`

dkirker replied:
I noticed this had been marked "resolved" but does not appear to be resolved?

lukhong replied:
fixed.

tpmanley commented on devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy:
```diff
@@ -0,0 +1,123 @@
+// Parse incoming device messages to generate events
+def parse(String description) {
+    def parseMap = zigbee.parseDescriptionAsMap(description)
+    log.info "LOG-INFO-Shade: In parse(), description:- ${description}"
```
What do you see logged here? I ask because it looks like multiple attributes must be reported even though we only configured one for attribute reporting. In order to make sure you process all the attributes you need to do something like https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/smartthings/smartsense-motion-sensor.src/smartsense-motion-sensor.groovy#L98.

lukhong replied:
fixed to log the only lift position lift attribute

lukhong committed:
fixed name as shadeLevel, removed space, update attribute parse

lukhong committed:
WWST 2487 Changed namespace to SmartThings

Signed-off-by: Ki-wook <kiwook.hong@samsung.com>

lukhong committed:
WWST-2487 Change the file path

Signed-off-by: Ki-wook <kiwook.hong@samsung.com>

greens commented on devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:
```diff
@@ -0,0 +1,138 @@
+    log.info "configure()"
+    sendEvent(name: "checkInterval", value: 2 * 60 * 60 + 2 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+    log.debug "Configuring Reporting and Bindings."
+    zigbee.configureReporting(CLUSTER_WINDOW_COVERING, ATTRIBUTE_POSITION_LIFT, DataType.UINT8, 0, 600, null)
```
Do we want a min reporting interval this low? I could see this behaving badly. Also I'm not sure how `null` has minReportableChange will work.

tpmanley replied:
0 means it will be reported whenever it changes subject to limitations of the device. I could see changing it to 1 which would mean report when it changes but no more often than 1/second. `null` isn't valid here

tpmanley commented on devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:
```diff
@@ -0,0 +1,138 @@
+
+        command "pause"
+
+        fingerprint profileId: "0104", inClusters: "0000, 0003, 0004, 0102", outClusters: "0019", model: "E2B0-KR000Z0-HA", deviceJoinName: "SOMFY Blind Controller/eZEX", vid:"SmartThings-smartthings-Somfy_WindowShade" // SY-IoT201-BD
```
nit: add spacer after `vid`. Also, could we just get metadata for this DTH instead of device specific metadata?

greens replied:
agree with tom that this vid should just be moved into the device definition

lukhong replied:
I thought that Somfy named vid is not good in the definition because we think this DTH as a generic zigbee window shade DTH.
Should I remove the vid from the fingerprint and move it to the definition with `generic-shade` or `SmartThings-smartthings-Somfy_WindowShade` ?

dkirker replied:
What they mean is instead of 

```
    definition(name: "ZigBee Window Shade", namespace: "smartthings", author: "SmartThings", ocfDeviceType: "oic.d.blind") {

...

        fingerprint profileId: "0104", inClusters: "0000, 0003, 0004, 0102", outClusters: "0019", model: "E2B0-KR000Z0-HA", deviceJoinName: "SOMFY Blind Controller/eZEX", vid:"SmartThings-smartthings-Somfy_WindowShade" // SY-IoT201-BD
```
to use
```
    definition(name: "ZigBee Window Shade", namespace: "smartthings", author: "SmartThings", ocfDeviceType: "oic.d.blind", mnmn: "SmartThings", vid: "generic-shade") {

...

        fingerprint profileId: "0104", inClusters: "0000, 0003, 0004, 0102", outClusters: "0019", model: "E2B0-KR000Z0-HA", deviceJoinName: "SOMFY Blind Controller/eZEX" // SY-IoT201-BD
```

lukhong replied:
Got it. Thank you :)

lukhong committed:
WWST-2487 Moved vid into definition with generic name

Signed-off-by: Ki-wook <kiwook.hong@samsung.com>

tpmanley commented on devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:
```diff
@@ -0,0 +1,138 @@
+        capability "Window Shade"
+        capability "Health Check"
+        capability "Switch Level"
+
```
How important is this pause command? Like, is it needed for this device to pass WWST?

lukhong replied:
Actually not. I think it is a useful command for the user but only available via SmartThings classic. I remained it because the classic app is still alive.

dkirker replied:
There is a capability which adds that feature. However, I am not sure yet if it is available for use.
